### PR TITLE
when a drive is closed cancel the cleanupTrash goroutine

### PR DESCRIPTION
**Pull request by harshavardhana**
## Community Contribution License
All community contributions in this pull request are licensed to the project maintainers
under the terms of the [Apache 2 license](https://www.apache.org/licenses/LICENSE-2.0). 
By creating this pull request I represent that I have the right to license the 
contributions to the project maintainers under the Apache 2 license.

## Description
when a drive is closed, cancel the cleanupTrash goroutine

## Motivation and Context
when a hung drive is hot-unplugged, the server might go 
into a loop where the previous `format.json` is somehow 
still accessible to the process, we try to re-init() drives, 
but that seems to cause a previous goroutine to hang around 
since it is not canceled away when the drive is closed.

Bonus: add deadline for immediate purge routine to unblock
it if the drive is blocking mutations.

## How to test this PR?
Use `xfs_freeze` to freeze the drive from mutations,

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Optimization (provides speedup with no functional changes)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Unit tests added/updated
- [ ] Internal documentation updated
- [ ] Create a documentation update request [here](https://github.com/minio/docs/issues/new?label=doc-change,title=Doc+Updated+Needed+For+PR+github.com%2fminio%2fminio%2fpull%2fNNNNN)

### CODE REVIEW

**Review by vadmeste — APPROVED** [review]

**Review by jiuker — APPROVED** [review]

**Review by shtripat — APPROVED** [review]
LGTM. Verified with xfs_freeze.
Once one mount is frozen, `mc mb` errors as below, but once unfrozen it goes thorugh
```
API: SYSTEM.storage
Time: 13:04:36 UTC 08/28/2024
DeploymentID: bc155795-d22e-468f-bbb9-b9c81430f937
Error: node(10.212.97.210:9000): taking drive /mnt/virtual-disk-1 offline: unable to write+read for 30.001s (*errors.errorString)
       4: internal/logger/logger.go:258:logger.LogAlwaysIf()
       3: cmd/logging.go:160:cmd.storageLogAlwaysIf()
       2: cmd/xl-storage-disk-id-check.go:1017:cmd.(*xlStorageDiskIDCheck).monitorDiskWritable.func1.1()
       1: cmd/xl-storage-disk-id-check.go:1037:cmd.(*xlStorageDiskIDCheck).monitorDiskWritable.func1.2()
```
Dont see any hanging and listing of existing buckets works fine during freeze.

File: cmd/xl-storage.go [hunk 1]
```diff
@@ -1202,7 +1208,12 @@ func (s *xlStorage) cleanupTrashImmediateCallers(ctx context.Context) {
 		case <-ctx.Done():
 			return
 		case entry := <-s.immediatePurge:
-			removeAll(entry)
+			// Add deadlines such that immediate purge is not
+			// perpetually hung here.
+			w := xioutil.NewDeadlineWorker(globalDriveConfig.GetMaxTimeout())
+			w.Run(func() error {
+				return removeAll(entry)
```
Comment (klauspost) [inline cmd/xl-storage.go:1215]: For a more timely response, don't we also want to propagate (and react to) a context?

Comment (klauspost) [reply cmd/xl-storage.go:1215]: hmmm.. I guess we'd need to fork the stdlib since this ends up in `os.RemoveAll()` :/

**Review by klauspost — APPROVED** [review]
Better than before :)

**Review by poornas — APPROVED** [review]

File: cmd/xl-storage.go (hunk 1 above)
Comment (harshavardhana) [reply cmd/xl-storage.go:1215]: yeah no way to do it

**Comment by harshavardhana** [comment]
> Dont see any hanging and listing of existing buckets works fine during freeze.

Hanging reproducing requires ILM to be configured with an automatic purge of the content, at that point you can see `rename()` and `unlink()` can potentially hang this goroutine.

**Pull request merged**